### PR TITLE
[Serve] [1/3 Wrappers] Allow `@serve.batch` to accept args and kwargs

### PR DESCRIPTION
**Pull request by simon-mo**
## Why are these changes needed?
This PR adds support for arbitrary number of args and kwargs for serve.batch decorator. This is needed for model wrapper integration because we relies on FastAPI's kwargs dependency injection to work.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/batching.py [hunk 1]
```diff
@@ -2,11 +2,43 @@
...
+
+def _batch_args_kwargs(
+    list_of_flatten_args: List[List[Any]],
+) -> Tuple[Tuple[Any], Dict[Any, Any]]:
+    """Batch a list of flatten args and returns regular args and kwargs"""
+    # Ray's flatten arg format is a list with alternating key and values
+    # e.g. args=(1, 2), kwargs={"key": "val"} got turned into
+    #      [None, 1, None, 2, "key", "val"]
+    arg_lengths = {len(args) for args in list_of_flatten_args}
+    assert (
```
Comment (edoakes) [inline python/ray/serve/batching.py:27]: this is just an internal assertion, not a user-facing error, right?

File: python/ray/serve/batching.py [hunk 2]
```diff
@@ -2,11 +2,43 @@
...
 
+@dataclass
+class SingleRequest:
+    self_arg: Optional[Any]
+    flat_args: List[Any]
+    future: asyncio.Future
+
+
+def _batch_args_kwargs(
+    list_of_flatten_args: List[List[Any]],
```
Comment (edoakes) [inline python/ray/serve/batching.py:20]:
```suggestion
    list_of_flattened_args: List[List[Any]],
```

File: python/ray/serve/batching.py [hunk 3]
```diff
@@ -2,11 +2,43 @@
...
+from dataclasses import dataclass
 
+from ray._private.signature import extract_signature, flatten_args, recover_args
 from ray.serve.exceptions import RayServeException
 
 
+@dataclass
+class SingleRequest:
+    self_arg: Optional[Any]
+    flat_args: List[Any]
```
Comment (edoakes) [inline python/ray/serve/batching.py:15]:
```suggestion
    flattened_args: List[Any]
```

use consistent naming across the code please, either choose "flattened_args" or "flat_args"

File: python/ray/serve/tests/test_batching.py [hunk 4]
```diff
@@ -259,6 +259,37 @@ async def call(arg):
...
+
+    if mode == "args":
+        coros = [func("hi1", "hi2"), func("hi3", "hi4")]
+    elif mode == "kwargs":
+        coros = [func(key1="hi1", key2="hi2"), func(key1="hi3", key2="hi4")]
+    elif mode == "mixed":
+        coros = [func("hi1", key2="hi2"), func("hi3", key2="hi4")]
+
+    result = await asyncio.gather(*coros)
+    assert result == [("hi1", "hi2"), ("hi3", "hi4")]
```
Comment (edoakes) [inline python/ray/serve/tests/test_batching.py:290]: please also test out-of-order kwargs (`arg2="hi2", arg1="hi1"`)

Comment (edoakes) [reply python/ray/serve/tests/test_batching.py:290]: none of the above tests verify that the logic is ordering-independent

File: python/ray/serve/batching.py (hunk 1 above)
Comment (simon-mo) [reply python/ray/serve/batching.py:27]: Correct.

**Comment by edoakes** [comment]
@simon-mo did you forget to push?

**Comment by simon-mo** [comment]
one sec, my merge is messed up

**Review by edoakes — APPROVED** [review]

**Comment by simon-mo** [comment]
@edoakes fixed

**Comment by edoakes** [comment]
No problem, I reviewed only the serve changes and LGTM. Please fix the conflicts before merging obviously :)

**Pull request merged**